mlkem/ntt.c

```diff
@@ -74,49 +74,95 @@ const int16_t zetas[128] = {
  **************************************************/
 #if !defined(MLKEM_USE_NATIVE_NTT)
 
-// Check that the specific bound for the reference NTT implies
-// the bound required by the C<->Native interface.
-#define NTT_BOUND_REF (5 * MLKEM_Q)
-STATIC_ASSERT(NTT_BOUND_REF <= NTT_BOUND, ntt_ref_bound)
+// Compute a block CT butterflies with a fixed twiddle factor
+STATIC_TESTABLE
+// clang-format off
+void ntt_butterfly_block(int16_t *r, int16_t root, int start, int len, int bound)
+  REQUIRES(0 <= start && start < MLKEM_N && 1 <= len && len <= MLKEM_N / 2 && start + 2 * len <= MLKEM_N)
+  REQUIRES(0 <= bound && bound < INT16_MAX - MLKEM_Q)
+  REQUIRES(IS_FRESH(r, sizeof(int16_t) * MLKEM_N))
+  REQUIRES(SCALAR_ABS_BOUND(root, HALF_Q - 1))
+  REQUIRES(ARRAY_ABS_BOUND(r, 0, start - 1, bound + MLKEM_Q))
+  REQUIRES(ARRAY_ABS_BOUND(r, start, MLKEM_N - 1, bound))
+  ASSIGNS(OBJECT_UPTO(r, sizeof(int16_t) * MLKEM_N))
+  ENSURES(ARRAY_ABS_BOUND(r, 0, start + 2*len - 1, bound + MLKEM_Q))
+  ENSURES(ARRAY_ABS_BOUND(r, start + 2 * len, MLKEM_N - 1, bound))
+// clang-format on
+{
+  // Parameter only used in the CBMC specification
+  ((void)bound);
+  // clang-format off
+  for (int j = start; j < start + len; j++)
+    ASSIGNS(j, OBJECT_UPTO(r, sizeof(int16_t) * MLKEM_N))
+    INVARIANT(start <= j && j <= start + len)
+    INVARIANT(ARRAY_ABS_BOUND(r, 0,           j - 1,           bound + MLKEM_Q))
+    INVARIANT(ARRAY_ABS_BOUND(r, j,           start + len - 1, bound))
+    INVARIANT(ARRAY_ABS_BOUND(r, start + len, j + len - 1,     bound + MLKEM_Q))
+    INVARIANT(ARRAY_ABS_BOUND(r, j + len,     MLKEM_N - 1,     bound))
+    // clang-format on
+    {
+      int16_t t;
+      t = fqmul(r[j + len], root);
+      r[j + len] = r[j] - t;
+      r[j] = r[j] + t;
+    }
+}
+
+// Compute one layer of forward NTT
+STATIC_TESTABLE
+void ntt_layer(poly *p, int layer)
+    // clang-format off
+  REQUIRES(IS_FRESH(p, sizeof(poly)))
+  REQUIRES(1 <= layer && layer <= 7)
+  REQUIRES(ARRAY_ABS_BOUND(p->coeffs, 0, MLKEM_N - 1, layer * MLKEM_Q))
+  ASSIGNS(OBJECT_UPTO(p, sizeof(poly)))
+  ENSURES(ARRAY_ABS_BOUND(p->coeffs, 0, MLKEM_N - 1, (layer + 1) * MLKEM_Q))
+// clang-format on
+{
+  int16_t *r = p->coeffs;
+  const int len = 1u << (8 - layer);
+  const int blocks = 1 << (layer - 1);
+  for (int i = 0; i < blocks; i++)
+      // clang-format off
+    ASSIGNS(i, OBJECT_UPTO(r, sizeof(poly)))
+    INVARIANT(1 <= layer && layer <= 7)
+    INVARIANT(0 <= i && i <= blocks)
+    INVARIANT(ARRAY_ABS_BOUND(r, 2 * i * len, MLKEM_N - 1, layer * MLKEM_Q))
```

Comment:
I would probably swap these two invariant over, so the ranges of the quanitifers are in
  0 .. 2 * i * len-1
followed by
  2 * i * len .. MLKEM_N-1
so it's obvious that the ranges "join up" and cover all the coefficients.

and possibly add explanatory comments as above.

Reply:
Does no longer apply